bin/tx-prover/src/proxy/mod.rs

```diff
@@ -54,15 +62,166 @@ impl LoadBalancer {
         self.available_workers.write().await.pop()
     }
 
-    /// Adds the provided worker to the list of available workers.
+    /// Adds a worker to the list of available workers.
     ///
-    /// # Panics
-    /// Panics if the provided worker is already in the list of available workers.
+    /// If the worker is already in the list of available workers, it won't be added.
+    /// If the worker is not in the list of current workers, it won't be added.
     pub async fn add_available_worker(&self, worker: Backend) {
         let mut available_workers = self.available_workers.write().await;
-        assert!(!available_workers.contains(&worker), "Worker already available");
-        info!("Worker {} is now available", worker.addr);
-        available_workers.push(worker);
+        let current_workers = self.current_workers.read().await;
+        if !available_workers.contains(&worker) && current_workers.contains(&worker) {
+            available_workers.push(worker);
+        }
+    }
+
+    /// Updates the list of available workers based on the given action ("add" or "remove").
+    ///
+    /// # Behavior
+    ///
+    /// ## Add Action
+    /// - If the worker exists in the current workers list, do nothing.
+    /// - If the worker does not exist in the current workers list, add it to both the current and
+    ///   available workers lists.
+    ///
+    /// ## Remove Action
+    /// - If the worker exists in the available workers list, remove it.
+    /// - If the worker exists in the current workers list, remove it.
+    /// - Otherwise, do nothing.
+    ///
+    /// Finally, updates the configuration file with the new list of workers.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - The worker address cannot be cast into [Backend].
+    /// - The configuration file cannot be loaded or saved.
+    /// - A worker address cannot be retrieved.
+    pub async fn update_workers(
+        &self,
+        update_workers: UpdateWorkers,
+    ) -> std::result::Result<(), String> {
+        let mut available_workers = self.available_workers.write().await;
+        let mut current_workers = self.current_workers.write().await;
```

Comment:
The fact that we have two separate locks for the separate worker lists may lead to a deadlock (I don't think it's the case here as the lock order is always the same). We may want to unify them so that we don't have that possibility.

Reply:
I could not see a situation on our particular code that leads to a deadlock, as you said, we always lock in the same order. Though if we prefer I can unify the locks.